Rakefile

```diff
@@ -1,10 +1,145 @@
+# frozen_string_literal: true
+
 require 'rake/extensiontask'
 require 'rspec/core/rake_task'
+require 'rake_compiler_dock'
+require 'yaml'
+
+CLEAN.include FileList['**/*{.o,.so,.dylib,.bundle}'],
+              FileList['**/extconf.h'],
+              FileList['**/Makefile'],
+              FileList['pkg/']
+
+CLOBBER.include FileList['**/tmp'],
+                FileList['**/*.log'],
+                FileList['doc/**'],
+                FileList['tmp/']
+CLOBBER.add("ports/*").exclude(%r{ports/archives$})
+
+RE2_GEM_SPEC = Gem::Specification.load('re2.gemspec')
+
+Gem::PackageTask.new(RE2_GEM_SPEC) do |p|
+  p.need_zip = false
+  p.need_tar = false
+end
+
+CROSS_RUBY_VERSIONS = %w[3.2.0 3.1.0 3.0.0 2.7.0].join(':')
+CROSS_RUBY_PLATFORMS = %w[
+  aarch64-linux
+  arm-linux
+  arm64-darwin
+  x64-mingw-ucrt
+  x64-mingw32
+  x86-linux
+  x86-mingw32
+  x86_64-darwin
+  x86_64-linux
+].freeze
+
+ENV['RUBY_CC_VERSION'] = CROSS_RUBY_VERSIONS
 
-Rake::ExtensionTask.new('re2')
+Rake::ExtensionTask.new('re2', RE2_GEM_SPEC) do |e|
+  e.cross_compile = true
+  e.cross_config_options << '--enable-cross-build'
+  e.config_options << '--disable-system-libraries'
+  e.cross_platform = CROSS_RUBY_PLATFORMS
+  e.cross_compiling do |spec|
+    spec.files.reject! { |path| File.fnmatch?('ports/*', path) }
+    spec.dependencies.reject! { |dep| dep.name == 'mini_portile2' }
+  end
+end
 
 RSpec::Core::RakeTask.new(:spec)
 
-task :spec    => :compile
-task :default => :spec
+namespace 'gem' do
+  def gem_builder(platform)
+    # use Task#invoke because the pkg/*gem task is defined at runtime
+    Rake::Task["native:#{platform}"].invoke
+    Rake::Task["pkg/#{RE2_GEM_SPEC.full_name}-#{Gem::Platform.new(platform)}.gem"].invoke
+  end
+
+  CROSS_RUBY_PLATFORMS.each do |platform|
+    # The Linux x86 image (ghcr.io/rake-compiler/rake-compiler-dock-image:1.3.0-mri-x86_64-linux)
+    # is based on CentOS 7 and has two versions of cmake installed:
+    # a 2.8 version in /usr/bin and a 3.25 in /usr/local/bin. The latter is needed by abseil.
+    cmake =
+      case platform
+      when 'x86_64-linux', 'x86-linux'
+        '/usr/local/bin/cmake'
+      else
+        'cmake'
+      end
```

Comment:
Would this work if `CMAKE` is an environment variable? If so, perhaps we could lift this into the workflow only for those images.

Reply:
I wanted to make it possible to build locally with `rake gem:native`, so setting `CMAKE` only in the CI job would prevent that.